libvast/src/system/index.cpp

```diff
@@ -41,24 +47,692 @@
+    auto partition_id = *it++;
+    if (auto partition_actor = spin_up(partition_id))
+      result.push_back(std::make_pair(partition_id, partition_actor));
+  }
+  lookup.partitions.erase(lookup.partitions.begin(), it);
+  VAST_DEBUG(self, "launched", result.size(),
+             "await handlers to fill the pending query map");
+  return result;
+}
+
+/// Sends an `evaluate` atom to all partition actors passed into this function,
+/// and collects the resulting
+/// @param c Continuation that takes a single argument of type
+/// `caf::expected<pending_query_map>`. The continuation will be called in the
+/// context of `self`.
+//
+// TODO: At some point we should add some more template magic on top of
+// this and turn it into a generic functions that maps
+//
+//   (map from U to A, request param pack R, result handler with param X) ->
+//   expected<map from U to X>
+template <typename Continuation>
+void await_evaluation_maps(
+  caf::stateful_actor<index_state>* self, const expression& expr,
+  const std::vector<std::pair<vast::uuid, caf::actor>>& actors,
+  Continuation c) {
+  struct counter {
+    size_t received;
+    pending_query_map pqm;
+  };
+  auto expected = actors.size();
+  auto shared_counter = std::make_shared<counter>();
+  for (auto& [id, actor] : actors) {
+    auto partition_id = id; // Can't use structured binding inside lambda.
+    self->request(actor, caf::infinite, expr)
+      .then(
+        [=](evaluation_triples triples) {
+          auto received = ++shared_counter->received;
+          if (!triples.empty())
+            shared_counter->pqm.emplace(partition_id, std::move(triples));
+          if (received == expected)
+            c(std::move(shared_counter->pqm));
+        },
+        [=](caf::error err) {
+          // Don't increase `received` to ensure the sucess handler never gets
+          // called.
+          c(err);
+        });
+  }
+}
+
+query_map
+index_state::launch_evaluators(pending_query_map& pqm, expression expr) {
+  query_map result;
+  for (auto& [id, eval] : pqm) {
+    std::vector<caf::actor> xs{self->spawn(evaluator, expr, std::move(eval))};
+    result.emplace(id, std::move(xs));
+  }
+  pqm.clear();
+  return result;
+}
+
+path index_state::index_filename(path basename) const {
+  return basename / dir / "index.bin";
+}
+
+caf::expected<flatbuffers::Offset<fbs::Index>>
+pack(flatbuffers::FlatBufferBuilder& builder, const index_state& state) {
+  auto meta_idx = pack(builder, state.meta_idx);
+  if (!meta_idx)
+    return meta_idx.error();
+  VAST_VERBOSE(state.self, "persisting", state.persisted_partitions.size(),
+               " definitely persisted and ", state.unpersisted.size(),
+               " maybe persisted partitions uuids");
+  std::vector<flatbuffers::Offset<fbs::UUID>> partition_offsets;
+  for (auto uuid : state.persisted_partitions) {
+    auto uuid_fb = pack(builder, uuid);
+    if (!uuid_fb)
+      return uuid_fb.error();
+    partition_offsets.push_back(*uuid_fb);
```

Comment:
```suggestion
    if (auto uuid_fb = pack(builder, uuid));
        partition_offsets.push_back(*uuid_fb);
    else
      return uuid_fb.error();
```

I find this `expected<T>` pattern more idiomatic, as it brings out the if-else character better. But it's a matter of preference.